### ISSUE TITLE
IOs are failing when snapshot is created out of an LVOL

### ISSUE THREAD

**Issue by SavanAnjaria** [spdk/spdk#603]
## Expected Behavior
When IOs are in progress on a LVOL, snapshot creation out of that LVOL shouldn't harm the ongoing IOs.

## Current Behavior
When IOs are in progress on a LVOL, snapshot creation out of that LVOL throws following error messages on the target side and fails the IOs.

bdev_nvme.c:1640:bdev_nvme_queue_cmd: *ERROR*: writev failed: rc = -22
nvme_pcie.c:1716:nvme_pcie_prp_list_append: *ERROR*: vtophys(0x28c01c8) failed
nvme_qpair.c: 137:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:180 nsid:1 lba:2153645032 len:8
nvme_qpair.c: 306:nvme_qpair_print_completion: *NOTICE*: INVALID FIELD (00/02) sqid:1 cid:180 cdw0:0 sqhd:0000 p:0 m:0 dnr:1

**fio config used**
[global]
bs=4k
iodepth=1
direct=1
ioengine=libaio
numjobs=1
randrepeat=0
group_reporting
time_based
runtime=60
filesize=20G

[job1]
rw=write
filename=/dev/nvme0n1 (this is the volume from which snapshot was created on the target side)

On debugging little further into this issue, I could see that we usually hit this issue when the frozen IOs (during snapshot creation) are completed and we allow the new IOs on that volume on already allocated clusters (is_page_allocated would return true).

## Possible Solution
1. Either some sort of buffer corruption for the buffer used to send the data to the drives. OR
2. Some issue with DPDK (based on the history of such issues reported). OR
3. Some restrictions of IO patterns (based on the configuration).

## Steps to Reproduce
Target Side:
1.	Create NVMe Bdev
2.	Construct logical volume store (LVS) on NVMe Bdev
3.	Construct logical volume (LVOL) on the LVS created
4.	Create subsystem and add listener to that of type RDMA

Initiator Side:
1.	Discover the logical volume
2.	Connect to the logical volume
3.	Start fio (writes only)

Target Side:
1.	Create a snapshot from the logical volume created earlier when the IOs are in progress

## Context (Environment including OS version, SPDK version, etc.)
SPDK Version: master branch based of Change-ID: I38e268b532ff3ce87721c02f15ac4f674856d103
Initiator OS Version: CentOS Linux 7 (Core)
Target OS Version: CentOS Linux 7 (Core)
[Snapshot_With_IO_Fails.txt](https://github.com/spdk/spdk/files/2784448/Snapshot_With_IO_Fails.txt)

**Comment by tomzawadzki** [comment]
Hi @SavanAnjaria 
Could you please submit logs from target which has debug enabled and displaying debug logs (./nvmf_tgt -L blob -L blob_rw -L lvol -L vbdev_lvol -L bdev) ?

**Comment by SavanAnjaria** [comment]
> Hi @SavanAnjaria
> Could you please submit logs from target which has debug enabled and displaying debug logs (./nvmf_tgt -L blob -L blob_rw -L lvol -L vbdev_lvol -L bdev) ?

[SPDK_Issue_603_Logs.zip](https://github.com/spdk/spdk/files/2821057/SPDK_Issue_603_Logs.zip)

@tomzawadzki : I have uploaded the logs as requested. Please look for following string in order to find the failure condition.

nvme_pcie.c:1716:nvme_pcie_prp_list_append: *ERROR*: vtophys(0x7f065efaa338) failed
nvme_qpair.c: 137:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:157 nsid:1 lba:134939000 len:8
nvme_qpair.c: 306:nvme_qpair_print_completion: *NOTICE*: INVALID FIELD (00/02) sqid:1 cid:157 cdw0:0 sqhd:0000 p:0 m:0 dnr:1

**Comment by SavanAnjaria** [comment]
@darsto : Can you please remove the tag Waiting on Submitter as I have already provided required details?

**Comment by SavanAnjaria** [comment]
@jimharris 
Sometime back, https://github.com/spdk/spdk/issues/226 was reported and that had similar signature for failure (nvme_pcie_prp_list_append: _ERROR_: vtophys(<address>) failed and https://review.gerrithub.io/c/spdk/spdk/+/396318 was the change made to address that issue. Do you see any similarity here between #226 and #603?

**Comment by SavanAnjaria** [comment]
@jimharris 
#274 also shows similar footprint. Just trying to give pointers in order to relate this issue and get quick help based on that.

**Comment by jimharris** [comment]
Thanks @SavanAnjaria.  @tomzawadzki will be back in the office on Monday - I'm sure he'll take a look at your pointers as he digs into this further.

**Comment by tomzawadzki** [comment]
@SavanAnjaria Thank you for submitting the logs. As you've pointed out it seems that the error occurs after the snapshot completes and writing is resumed after 'freeze_io'.

I've been trying to reproduce this without using NVMe-oF without success so far. With only NVMe device I had no trouble with snapshotting during FIO submitting IO. There is functional test within autotest for this (test_cases.py 752).

This might be related to interaction with NVMe-oF, but spdk_vtophys() failing is unusual and does not seem related to the lvol/blobstore. @SavanAnjaria Have you seen this without lvols present ?

I'll try to setup NVMe-oF for testing this, and will update with further findings.

**Comment by SavanAnjaria** [comment]
@tomzawadzki  Thanks for looking into this. Initially I had also thought that, this is related to NVMe-oF. But, later I found that the following prints are generated from nvme_pcie.c which is responsible for transferring the data to the NVMe Drives I believe.
**nvme_pcie.c**:1716:nvme_pcie_prp_list_append: _ERROR_: vtophys(0x28c01c8) failed

Earlier when such issues were reported (not related to snapshot but, had vtophys failed), either there was a change suggested in DPDK (#226 https://review.gerrithub.io/c/spdk/spdk/+/396318) or some suggestions were given to change the IO pattern/size.

So, I was wondering if for the present issue also, we are landing into the same zone.

**Comment by SavanAnjaria** [comment]
@tomzawadzki I haven't seen this issue without lvols being present. This was primarily seen with snapshot creation when IOs were in progress on the lvol (from which the snapshot was created).

**Comment by SavanAnjaria** [comment]
@tomzawadzki In the unit test 752, we are setting Thin provision property of the lvol prior to creating snapshot. My understanding was, the thin provision property of the lvol gets set when we create the snapshot.

Moreover when we create snapshot out of an lvol, the clone property of an lvol gets set. I was curious as to why do we set clone property of lvol on snapshot creation.

**Comment by tomzawadzki** [comment]
@SavanAnjaria This is unrelated to the issue, but please let me explain the reasoning behind some of your observations.

Thin provisioning feature is not directly related to snapshot/clone functionality. An lvol can be created with thin provisioning set by providing '-t' argument in RPC.

Although there is additional caveat for clones. To make them work as they do currently, snapshot is an lvol that is set to read only (can be thin or thick provisioned). Meanwhile clone is an lvol that is thin provisioned by default and has snapshot set as backing bs_dev.

The snapshotting process does look quite complex under the hood. The assumption was that when user creates snapshot - a new bdev set to read only is presented, meanwhile old bdev remains RW as previously. This results in following operations on creating snapshot:
- create new blob that will be the snapshot
- set freeze_io on the original blob
- copy properties and cluster map from original blob to new one
- change properties (set thin provision and backing bs_dev) of original blob and zero out the cluster map
- unfreeze io on the original blob

This allows the user to use the original blob (now a clone) as it was used previously.

**Comment by SavanAnjaria** [comment]
@tomzawadzki Understood. Thanks for the explanation

**Comment by tomzawadzki** [comment]
@SavanAnjaria I was successful in getting a reproduction of your issue with NVMe-oF setup. This happens just after snapshot is finished and on one of writes after new cluster allocation.

For reference log attached is snippet of logs between blob sync at snapshot finish and the vtophys error.
[log_final.txt](https://github.com/spdk/spdk/files/2889917/log_final.txt)

The virt_addr does not look right at all for this case, looking further into this.

**Comment by SavanAnjaria** [comment]
@tomzawadzki That's great. Yes, virt_address doesn't seem correct in the logs you have provided.

**Comment by darsto** [comment]
Bug scrub: @jimharris to provide TOmek some suggestions what to debug

**Comment by jimharris** [comment]
@tomzawadzki - can you set a breakpoint when vtophys fails, and then look at the backtrace?

Does this happen on the first thin-provisioned cluster allocation?
Would it be possible to post more of the log on a failure case?

**Comment by SavanAnjaria** [comment]
@tomzawadzki @jimharris 
I could reproduce this issue with another use case as well
1. Create NVMe Bdev
2. Create LVS on that
3. Create a Thin Provisioned LVOL on the LVS
4. Create subsystem, add this lvol to the subsystem and expose that to the initiator
5. Start Writes on the Thin Provisioned LVOL from the Initiator
6. Do resize_lvol_bdev when Writes are in progress

Thought of providing this additional info as this looks to be a more straight forward scenario. Common things between the earlier scenario and this one are:
1. LVOL is thin provisioned (on which IOs are running)
2. We are freezing and unfreezing IOs

**Comment by darsto** [comment]
@tomzawadzki Please do some scoping for this. Can we fix it for 19.04?

**Comment by darsto** [comment]
Bug scrub: try to reproduce this with only a single cpu

**Comment by benlwalker** [comment]
I wrote a simple test that creates a thin provisioned lvol, does some writes to it, and while those writes are happening it resizes the lvol larger. If I run this with a target using just 1 core, it works fine. If I give the target 4 cores, address sanitizer detects a use after free every time. There is some bug in allocating clusters from multiple threads occurring. We should be able to get to a root cause shortly.

**Comment by tomzawadzki** [comment]
- The tests were performed with **single CPU core**, so this occurs even with single CPU core. [snapshot_repro.tar.gz](https://github.com/spdk/spdk/files/3118387/snapshot_repro.tar.gz)

- @SavanAnjaria Tested with resize scenario as well. It occurs but much less frequent. IO hits the freeze io window much less often here. Please see attachment 
[resize_repro.tar.gz](https://github.com/spdk/spdk/files/3118385/resize_repro.tar.gz) for log with multiple resizes on a **thick** provisioned, actively used lvol.

- As of current master, the issue described in this report additionally results in hitting assert(bdev_ch->io_outstanding > 0); in bdev layer.

- As a remainder reproduction steps didn't work on NVMe bdev or iSCSI target. Now in addition verified that it does not reproduce with malloc bdev exposed via NVMe-oF target. So far couldn't find any relation in errors/fails to the NVMe bdev + NVMe-oF target combination. Might need to go back and check if the allocation occurs during freeze io in scenarios that do not reproduce.

Steps that the failing IO go through when **snapshotting** are:
1) Original submission is stopped by freeze_refcnt (Freeze IO in progress)
2) IO is resubmitted as part of unfreezing process, but hits unallocated cluster (due to 'old' blob being now a thin_provisioned). Put yet again on hold after the cluster allocation.
3) As the cluster finishes allocation, the IO is about to be submitted to the device. During which the incorrect vtophys translation occurs.

For **resize** steps are:
1) Original submission is stopped by freeze_refcnt (Freeze IO in progress)
2) IO is resubmitted as part of unfreezing process and hits allocated cluster (because blob thick provisioned). As IO do device is submitted the incorrect vtophys translation occurs.
Note: In this scenario resize claims clusters, but no thin provisioning is present. The newly claimed clusters are not accessible via NVMe-oF since there is no notification back to the initiator regarding change in size. So IO does not hit new clusters.

Not sure why would freezing io cause the vtophys hit following path in spdk_mem_map_translate()
```
vfn_2mb = vaddr >> SHIFT_2MB;
idx_256tb = MAP_256TB_IDX(vfn_2mb);
idx_1gb = MAP_1GB_IDX(vfn_2mb);

map_1gb = map->map_256tb.map[idx_256tb];
if (spdk_unlikely(!map_1gb)) {
	->>>> return map->default_translation;
}
```

**Comment by benlwalker** [comment]
Here's a test that reproduces the problem:

https://review.gerrithub.io/c/spdk/spdk/+/452150

You have to change the NVMe device BDF to whatever it is on your system.

**Comment by jimharris** [comment]
This patch fixes the issue on my system using Ben's test.  Go forth and test.

https://review.gerrithub.io/c/spdk/spdk/+/452157

**Comment by tomzawadzki** [comment]
Confirming that this fixes the issue on my system as well.

**Comment by darsto** [comment]
fix merged.
